internal/codespaces/grpc/client.go

```diff
@@ -0,0 +1,130 @@
+package grpc
+
+// gRPC client implementation to be able to connect to the gRPC server and perform the following operations:
+// - Start a remote JupyterLab server
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/cli/cli/v2/internal/codespaces/grpc/jupyter"
+	"github.com/cli/cli/v2/pkg/liveshare"
+	"golang.org/x/crypto/ssh"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	connectionTimeout = 5 * time.Second
+	requestTimeout    = 30 * time.Second
+)
+
+const (
+	codespacesInternalPort        = 16634
+	codespacesInternalSessionName = "CodespacesInternal"
+)
+
+type Client struct {
+	conn          *grpc.ClientConn
+	token         string
+	listener      net.Listener
+	jupyterClient jupyter.JupyterServerHostClient
+}
+
+type liveshareSession interface {
+	KeepAlive(string)
+	OpenStreamingChannel(context.Context, liveshare.ChannelID) (ssh.Channel, error)
+	StartSharing(context.Context, string, int) (liveshare.ChannelID, error)
+}
+
+// Finds a free port to listen on and creates a new gRPC client that connects to that port
+func Connect(ctx context.Context, session liveshareSession, token string) (*Client, error) {
+	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", 0))
+	if err != nil {
+		return nil, fmt.Errorf("failed to listen to local port over tcp: %w", err)
+	}
+
+	// Tunnel the remote gRPC server port to the local port
+	localPort := listener.Addr().(*net.TCPAddr).Port
+	internalTunnelClosed := make(chan error, 1)
+	go func() {
+		fwd := liveshare.NewPortForwarder(session, codespacesInternalSessionName, codespacesInternalPort, true)
+		internalTunnelClosed <- fwd.ForwardToListener(ctx, listener)
+	}()
+
+	// Create the gRPC client
+	client, err := NewClient(ctx, session, token, localPort)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create client: %w", err)
+	}
+
+	// Attach the listener so we can close it later
+	client.listener = listener
+
+	return client, err
+}
+
+// Creates a new gRPC client that connects to the given port
+func NewClient(ctx context.Context, session liveshareSession, token string, localPort int) (*Client, error) {
```

Comment:
Why does this method exist or need to be public?

Reply:
Also, passing context into a constructor is an anti-pattern in Go. See: https://github.com/golang/go/wiki/CodeReviewComments#contexts

>Don't add a Context member to a struct type; instead add a ctx parameter to each method on that type that needs to pass it along. The one exception is for methods whose signature must match an interface in the standard library or in a third party library.

I understand you need it to dial, but this usually means the constructor is doing too much, perhaps we need a separate dial function that takes a ctx parameter.

Reply:
For testing, we have the mock gRPC server running on a static port so we don't need these lines (and it panics on line 56 since we don't have a mock for the port forwarder) https://github.com/dmgardiner25/cli/blob/f947020fa5b957b89d4f7be200e70bf2f2a47b48/internal/codespaces/grpc/client.go#L46-L57 and having a separate `NewClient` function seemed like a good way to skip it when we know the exact port we want to connect to. To be honest, it is test-specific code but I can't think of a better way to test this. Open to any ideas you have!

Reply:
I think you need a mock liveshare session object that will connect you to the running grpc server. You already interfaced the session, which is good, but now the tests should provide it with a mocked session.

If you look at the three methods that we use in the session, the most important one to mock will be:
```go
OpenStreamingChannel(context.Context, liveshare.ChannelID) (ssh.Channel, error)
```

Ultimately, you can implement a mock `ssh.Channel` since it's also just an interface that has a Read/Write method (amongst other things, that you can stub). https://pkg.go.dev/golang.org/x/crypto/ssh#Channel

Note: I know what I'm suggesting is not super easy, but it is necessary for these network operations. You can see code in `port_forwarder_test.go` and `liveshare/test/server.go`, the type of code that supports our current implementation. I think as we move into this new grpc world and basis, these types of tests will be necessary and helpful, so working towards building up a way to do this will pay off with our test endpoints that are coming.

Reply:
Also, happy to schedule a pairing session sometime this week if you want!

Reply:
That would be great! Feel free to make a meeting for whenever works for you.